### ISSUE TITLE
initial prototype of React/Typescript frontend

### ISSUE THREAD

**Issue by grahame** [ealgis/ealgis#20]
get a environment going with a simple list of maps displayed (via API)

this requires

 - webpack
 - typescript
 - react
 - redux
 - react-router-redux
 - materialui

on landing on the page, the frontend should check if it is logged in. if it isn't, go to a login screen. if it isn't, go to a list of maps with a topbar showing info on the currently logged in user

**Comment by keithamoss** [comment]
A quick review of functionality and components on EAlGIS 1.0 that we'll want to port to 2.0.

- [ ] About dialog/page
- [ ] Map List (Choose map/Switch to a new map)
- [ ] Map Form GUI
  - [ ] Create map
  - [ ] Edit map
  - [ ] Delete map
- [ ] Layer List
- [ ] Layer Form GUI
  - [ ] Create layer
  - [ ] Edit layer
  - [ ] Clone layer (?)
  - [ ] Delete layer

**Comment by keithamoss** [comment]
Documenting the Great TypeScript/React/Redux/Redux-Thunk/Async Rabbit Hole.

**tl;dr** This ecosystem is horrendously immature, has nothing approaching consensus on best practice, and is so fragmented that any documentation and educational material around it is so poor from a bigger picture POV as to be worthless.

/rant

# Example Apps
## React + Redux + Router + Thunk
[favesound-redux](https://github.com/rwieruch/favesound-redux)

**Summary:** Complex app that covers must bases. Alas, not TypeScript.

## React + Redux
[Official Redux Async Example](https://github.com/reactjs/redux/tree/master/examples/async)
[Tutorial Writeup](http://redux.js.org/docs/advanced/AsyncFlow.html)

**Summary:** Simple example of Ajaxing using Redux. Alas, not TypeScript and no Thunk.


# Discussion Threads

**Comment by keithamoss** [comment]
# (ノಠ益ಠ)ノ彡┻━┻

**Comment by keithamoss** [comment]
I've curated a bit of a collection of "React ecosystem for beginners resources".

Link spam for now, these just need a bit of sorting into categories.

http://blog.andrewray.me/reactjs-for-stupid-people/
http://blog.andrewray.me/flux-for-stupid-people

http://blog.andrewray.me/reactjs-for-stupid-people/

http://www.jomendez.com/2017/01/26/redux-with-react-first-look/

http://blog.isquaredsoftware.com/2016/12/practical-redux-part-5-loading-and-displaying-data/#adding-a-mock-api-for-sample-data

https://zapier.com/engineering/react-js-tutorial-guide-gotchas/

https://online.reacttraining.com/p/reactjsfundamentals

https://github.com/ctrlplusb/react-universally

https://github.com/ctrlplusb/react-async-component

https://medium.com/@craigcartmell/complex-react-apps-and-the-many-pitfalls-part-1-of-infinity-fab2b2ccba57#.1pdy67e5f

https://facebook.github.io/react/blog/2016/07/22/create-apps-with-no-configuration.html

http://andrewhfarmer.com/react-ajax-best-practices/

https://egghead.io/courses/building-react-applications-with-idiomatic-redux

https://egghead.io/courses/getting-started-with-redux

http://teropa.info/blog/2015/09/10/full-stack-redux-tutorial.html

https://medium.com/audelabs/notes-from-my-first-react-redux-project-3f799beeb140#.uizjt8utt

**Comment by keithamoss** [comment]
Our journey migrating 100k lines of code from AngularJS to React (Chapter 1)
https://www.reddit.com/r/reactjs/comments/5qh60r/our_journey_migrating_100k_lines_of_code_from/

https://egghead.io/courses/react-fundamentals

**Comment by keithamoss** [comment]
**On React and OpenLayers**
[ol-react](https://github.com/richardhills/ol-react) _(Wot we're using.)_

Alternatives & Research:
[ol3-react-example](https://github.com/pka/ol3-react-example)
[react-open-layers](https://github.com/changey/react-open-layers)
[Examples of React js with Openlayers?](https://discuss.reactjs.org/t/examples-of-react-js-with-openlayers/867)
[How to use OpenLayers with react-starter-kit?](http://stackoverflow.com/questions/32422463/how-to-use-openlayers-with-react-starter-kit)

**Comment by keithamoss** [comment]
**On Material-UI and Drawers**
There's no inbuilt Drawer-AppBar layout as per the usual Material Design approach.

Ended up using flexbox instead: https://philipwalton.github.io/solved-by-flexbox/demos/holy-grail/

Research:
http://ramblings.mcpher.com/Home/excelquirks/reactredux/appnavdrawers
http://stackoverflow.com/questions/28691964/material-ui-open-leftnav-drawer-on-appbar-click
http://geon.github.io/programming/2016/02/24/flexbox-full-page-web-app-layout

**Comment by keithamoss** [comment]
**On React and The Use of Navigation Bars**
https://reacttraining.com/react-router/examples/sidebar
https://gist.github.com/CalebEverett/712fb3948d2695a95957
http://codetheory.in/react-integrating-routing-to-material-uis-left-nav-or-other-components/

**Comment by keithamoss** [comment]
**Re: Go to origin -**

The issue with this is that `ol-react` limits updates to the `ol.View` while the component is mounting, so any changes we make to the props never make it past `updateFromProps_()` in `view.js`.

I think `ol-react` does this to avoid infinite recursion...

While we can get at the internals of `olr.View` [thanks to refs](http://andrewhfarmer.com/how-to-pass-events-between-components/), I'm lost at how we could [invoke a call to `ol.view.animate()`](http://openlayers.org/en/latest/examples/animation.html) from MapUINavContainer.

As I understand it, in a perfect React-Redux world all we'd do is have our other component change the props that `olr.View` is bound to (in our case:`maps.map.json.map_defaults`), and that would flow through.

Needs some further thought.

**Comment by keithamoss** [comment]
And we're done!
